### ISSUE TITLE
Investigate taar and taar-lite powered API endpoints performance

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#16063]
This is probably out of our control, but taar and taar-lite powered API endpoints are considerably slower than they used to be:
![Screenshot_2020-11-30 AMO Prod frontend APIs usage performance - Grafana(2)](https://user-images.githubusercontent.com/187006/100602685-8342a200-3304-11eb-86f9-5c3713a2fa8f.png)
![Screenshot_2020-11-30 AMO Prod frontend APIs usage performance - Grafana(1)](https://user-images.githubusercontent.com/187006/100602686-83db3880-3304-11eb-95e6-34f0e8f1349f.png)

This is probably a combination of their migration to GCP and/or changes on their side, but we should investigate to find out if there is something we did that caused this, and what we could do to improve performance regardless of the cause.

**Comment by willdurand** [comment]
I am wondering if those endpoints are slower because they "fallback" more often than before, WDYT? We've seen lots of errors coming from the TAAR service in Sentry over the last few months.

**Comment by diox** [comment]
Yes, that's my theory as well. It's likely we're erroring more or just waiting for a response until the timeout more than before, and that causes the slowdown. See also https://bugzilla.mozilla.org/show_bug.cgi?id=1668614 which could be related.

**Comment by eviljeff** [comment]
we could reduce the timeout to increase our API performance? (at the expense of getting even more timeouts)

**Comment by willdurand** [comment]
Shouldn't we increase the timeout instead so that we give more time to TAAR to reply? (which should still be faster than our fallback code?)

**Comment by diox** [comment]
Looking more closely at Sentry, the timeout is rarely reached now - it used to be worse. It's set to one second so it doesn't really matter here I think - the slowness is probably for all requests, even successful ones. It might be because of the fallback but I suspect this is more on the taar side, partly because taar is on GCP and we're not.

**Comment by diox** [comment]
I created [a dashboard to monitor performance of all "external" services in AMO](https://earthangel-b40313e5.influxcloud.net/d/SEUKky0Gz/amo-external-services-perf?orgId=1). It doesn't distinguish between taar and taar lite (we use the same statsd timer) but it clearly shows perf getting worse in the end of September:
![Screenshot_2020-12-03 AMO External Services Perf - Grafana](https://user-images.githubusercontent.com/187006/101018739-35c86e00-356c-11eb-80c4-0e953f0834de.png)

**Comment by diox** [comment]
Perf hit does coincide with [the date we switched TAAR to GCP](https://bugzilla.mozilla.org/show_bug.cgi?id=1651393#c27) (09/24). That cost us ~200ms per call, which is not great, but I still can't explain the second spike in the `DiscoveryViewSet` graph. It doesn't show up on the graph that monitors the requests to TAAR, and doesn't show up on [TAAR graphs themselves](https://earthangel-b40313e5.influxcloud.net/d/riQqJonMk/taar?orgId=1&from=now-30d&to=now)

**Comment by diox** [comment]
In case this does turn out to be related to the errors we've been seeing somehow, here is the bug we filed a while back about them: https://bugzilla.mozilla.org/show_bug.cgi?id=1668614

**Comment by eviljeff** [comment]
No changes to `DiscoveryViewSet` (and no other obvious changes either) in https://github.com/mozilla/addons-server/compare/2020.10.22-1...2020.10.29 - assuming that's the date on the chart for the spike.

**Comment by diox** [comment]
Revisiting this issue with some profiling data on dev.

We can see that for both the call to taar-lite and taar... it's happening twice! That's because we're calling `https://<host>/<prefix>/<guid>` - no trailing slash - and it redirects to `https://<host>/<prefix>/<guid>/` with the trailing slash... So time it takes to get the answer is doubled everytime...

**Comment by diox** [comment]
Actually the double request is just for taar-lite - for taar, we do a `POST` to an URL that does have a trailing slash. This request does take a while though, up to 465ms on dev...

**Comment by diox** [comment]
See also https://github.com/mozilla/addons-server/issues/16182

**Comment by diox** [comment]
See also https://github.com/mozilla/taar/issues/113

**Comment by KevinMind** [comment]
Old Jira Ticket: https://mozilla-hub.atlassian.net/browse/ADDSRV-48
